### PR TITLE
support run work agent outside of managed cluster

### PR DESCRIPTION
**Pull request by zhujian7**
Signed-off-by: zhujian <jiazhu@redhat.com>

related issue: https://github.com/open-cluster-management-io/registration-operator/issues/158

---
todo:
- [ ] add description and script to deploy work agent in Hosted mode. (not contained in this PR)

### CODE REVIEW

File: pkg/spoke/spokeagent.go [hunk 1]
```diff
@@ -43,6 +46,7 @@ func (o *WorkloadAgentOptions) AddFlags(cmd *cobra.Command) {
 	flags := cmd.Flags()
 	// This command only supports reading from config
 	flags.StringVar(&o.HubKubeconfigFile, "hub-kubeconfig", o.HubKubeconfigFile, "Location of kubeconfig file to connect to hub cluster.")
+	flags.StringVar(&o.SpokeKubeconfigFile, "spoke-kubeconfig", o.SpokeKubeconfigFile, "Location of kubeconfig file to connect to spoke cluster.")
```
Comment (qiujian16) [inline pkg/spoke/spokeagent.go]: can't we use the built-in --kubeconfig flag here?

Comment (zhujian7) [reply pkg/spoke/spokeagent.go]: I  think it is better not, the built-in `--kubeconfig` in our case represents the cluster where the work-agent pod run.

File: pkg/spoke/spokeagent.go [hunk 2]
```diff
@@ -24,10 +26,12 @@ import (
 
 // WorkloadAgentOptions defines the flags for workload agent
 type WorkloadAgentOptions struct {
-	HubKubeconfigFile string
-	SpokeClusterName  string
-	QPS               float32
-	Burst             int
+	HubKubeconfigFile   string
+	SpokeKubeconfigFile string
+	SpokeKubeconfig     *rest.Config
```
Comment (qiujian16) [inline pkg/spoke/spokeagent.go]: why do we need this?

File: pkg/spoke/spokeagent.go [hunk 3]
```diff
@@ -64,28 +69,39 @@ func (o *WorkloadAgentOptions) RunWorkloadAgent(ctx context.Context, controllerC
...
 	workInformerFactory := workinformers.NewSharedInformerFactoryWithOptions(hubWorkClient, 5*time.Minute, workinformers.WithNamespace(o.SpokeClusterName))
 
-	// Build dynamic client and informer for spoke cluster
-	spokeRestConfig := controllerContext.KubeConfig
-	spokeRestConfig.QPS = o.QPS
-	spokeRestConfig.Burst = o.Burst
-	spokeDynamicClient, err := dynamic.NewForConfig(spokeRestConfig)
+	// load spoke client config and create spoke clients,
+	// the work agent may running not in the spoke/managed cluster.
+	if o.SpokeKubeconfig == nil {
```
Comment (qiujian16) [inline pkg/spoke/spokeagent.go]: I do not see how it can be not nil

File: pkg/spoke/spokeagent.go (hunk 1 above)
Comment (qiujian16) [reply pkg/spoke/spokeagent.go]: this should be explicit callout that if this is not set, but --kubeconfig is set. kubeconfig will be used to connect to spoke.

File: pkg/spoke/spokeagent.go [hunk 4]
```diff
@@ -64,28 +69,39 @@ func (o *WorkloadAgentOptions) RunWorkloadAgent(ctx context.Context, controllerC
...
+		if o.SpokeKubeconfigFile != "" {
+			o.SpokeKubeconfig, err = clientcmd.BuildConfigFromFlags("" /* leave masterurl as empty */, o.SpokeKubeconfigFile)
+			if err != nil {
+				return fmt.Errorf("unable to load spoke kubeconfig from file %q: %w", o.SpokeKubeconfigFile, err)
```
Comment (qiujian16) [inline pkg/spoke/spokeagent.go]: I would make it a func

```go
 func (o *WorkloadAgentOptions) spokeKubeConfig(controllerContext) *rest.Config, error {
    if o.SpokeKubeconfigFile != "" {
        return clientcmd.BuildConfigFromFlags("" /* leave masterurl as empty */, o.SpokeKubeconfigFile)
    }

   return controllerContext.KubeConfig
}
```

File: pkg/spoke/spokeagent.go (hunk 2 above)
Comment (zhujian7) [reply pkg/spoke/spokeagent.go]: This was originally used for integration test, but now I think it can be deleted

**Review by qiujian16 — COMMENTED** [review]
/approve
/lgtm

**Pull request merged**